etrago/tools/execute.py

```diff
@@ -373,20 +374,74 @@ def drop_foreign_components(network):
 
         return foreign_bus, foreign_comp, foreign_series
 
+    def import_gen_from_links(network):
+        """
+        create gas generators from links in order to not lose them when
+        dropping non-electric carriers
+        """
+        gas_to_add = network.links[
+            network.links.carrier.isin(
+                [
+                    "central_gas_CHP",
+                    "OCGT",
+                    "H2_to_power",
+                    "industrial_gas_CHP",
+                ]
+            )
+        ].copy()
+        
+        # Drop generators from the links tables
+        network.links.drop(gas_to_add.index, inplace=True)
+        
+        for df in network.links_t:
+            if not network.links_t[df].empty:
+                network.links_t[df].drop(columns=gas_to_add.index,
+                                         inplace=True, errors= "ignore")
+        
+        gas_to_add.rename(columns={"bus1": "bus"}, inplace=True)
+        
+        # Discard all generators < 1kW
+        gas_to_add = gas_to_add[gas_to_add["p_nom"] >= 0.001]
+        
+        gas_to_add["Generator"] = gas_to_add["bus"] + " " + gas_to_add.index + gas_to_add["carrier"]
+        gas_to_add.set_index("Generator", drop = True, inplace= True)
+        gas_to_add = gas_to_add[gas_to_add.columns[gas_to_add.columns.isin(network.generators.columns)]]
+        
+        network.import_components_from_dataframe(gas_to_add, "Generator")
+
+        # Dealing with generators_t
+        columns_new = network.links_t.p0.columns[
+            network.links_t.p0.columns.isin(gas_to_add.index)]
+        
+        new_gen_t = network.links_t.p0[columns_new]
+        
+        network.generators_t.p = network.generators_t.p.join(new_gen_t)
```

Comment:
The p_nom and p0 of links are defined as installed capacity and dispatch from bus0 to bus1. In case of an OCGT this means for example that this is the gas usage. The outgoing electricity is smaller (efficiency < 1). So I think you should either multiply p0 by the efficiency or directly use` -1 * p1` (which is in the end the same).

Reply:
> The p_nom and p0 of links are defined as installed capacity and dispatch from bus0 to bus1. In case of an OCGT this means for example that this is the gas usage. The outgoing electricity is smaller (efficiency < 1). So I think you should either multiply p0 by the efficiency or directly use` -1 * p1` (which is in the end the same).

Right. network.links_t.p1 * -1 will be used instead.

Reply:
Thank you!